### PR TITLE
[SPARK-33024][SQL] Fix CodeGen fallback issue of UDFSuite in Scala 2.13

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
After `SPARK-32851` set `CODEGEN_FACTORY_MODE` to `CODEGEN_ONLY` of `sparkConf` in `SharedSparkSessionBase`  to construction `SparkSession`  in test, the test suite `SPARK-32459: UDF should not fail on WrappedArray` in s.sql.UDFSuite exposed a codegen fallback issue in Scala 2.13 as follow:

```
- SPARK-32459: UDF should not fail on WrappedArray *** FAILED ***
Caused by: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 47, Column 99: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 47, Column 99: No applicable constructor/method found for zero actual parameters; candidates are: "public scala.collection.mutable.Builder scala.collection.mutable.ArraySeq$.newBuilder(java.lang.Object)", "public scala.collection.mutable.Builder scala.collection.mutable.ArraySeq$.newBuilder(scala.reflect.ClassTag)", "public abstract scala.collection.mutable.Builder scala.collection.EvidenceIterableFactory.newBuilder(java.lang.Object)"
```

The root cause is `WrappedArray` represent `mutable.ArraySeq`  in Scala 2.13 and has a different constructor of `newBuilder` method.

The main change of is pr is add Scala 2.13 only code part to deal with  `case match WrappedArray` in Scala 2.13.
 
### Why are the changes needed?
We need to support a Scala 2.13 build

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Scala 2.12: Pass the Jenkins or GitHub Action

- Scala 2.13: All tests passed.

Do the following:

```
dev/change-scala-version.sh 2.13
mvn clean install -DskipTests  -pl sql/core -Pscala-2.13 -am
mvn test -pl sql/core -Pscala-2.13
```

**Before**
```
Tests: succeeded 8540, failed 1, canceled 1, ignored 52, pending 0
*** 1 TEST FAILED ***

```

**After**

```
Tests: succeeded 8541, failed 0, canceled 1, ignored 52, pending 0
All tests passed.
```

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @srowen a new problem has been found of `sql/core` module in Scala 2.13

File: sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/objects/MapObjectsCodeGenHelper.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions.objects
+
+import scala.collection.mutable.{Builder, IndexedSeq, WrappedArray}
+
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenContext
+
+object MapObjectsCodeGenHelper {
```
Comment (srowen) [inline sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/objects/MapObjectsCodeGenHelper.scala:23]: Is it really necessary to have separate copies of this class? Perhaps just looking at the scala version or other properties of the byte code would let you decide on one or the other?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/objects/MapObjectsCodeGenHelper.scala:23]: I think it is workable if we accept the use `Properties.versionNumberString.startsWith ("2.13")` condition to add `if else branch`branch, I will changed the code to this style first

Comment (LuciferYang) [reply sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/objects/MapObjectsCodeGenHelper.scala]: Address 9c36f6c fix this comments, use `Properties.versionNumberString` to decide how to generate code.

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
This approach looks OK to me.

**Comment by LuciferYang** [comment]
@srowen now Jenkins test passed

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
thx ~ @srowen

**Pull request closed without merging**